File: rpg-backend/app/schemas/chest.py
from pydantic import BaseModel

from app.schemas.item import ItemAffixOut, ItemStatsOut


class ChestRarityProbabilityOut(BaseModel):
    rarity: str
    probability: float


class ChestOut(BaseModel):
    id: int
    slug: str
    name: str
    description: str
    price: int
    image_path: str | None
    guaranteed_min_rarity: str | None
    is_active: bool
    rarity_probabilities: list[ChestRarityProbabilityOut]


class ChestSummaryOut(BaseModel):
    id: int
    name: str


class ChestRewardOut(BaseModel):
    item_id: int
    item_template_id: int
    name: str
    slot: str
    tier: int
    rarity: str
    image_path: str | None
    stats: ItemStatsOut
    affixes: list[ItemAffixOut]


class ChestOpenResult(BaseModel):
    opening_id: int
    chest: ChestSummaryOut
    reward: ChestRewardOut
    balance: int


class OpenChestRequest(BaseModel):
    idempotency_key: str | None = None


class FreeChestStatusOut(BaseModel):
    chest: ChestOut
    is_available: bool
    next_available_at: str | None


class ChestOpeningHistoryOut(BaseModel):
    id: int
    chest: ChestSummaryOut
    reward_item_id: int
    reward_item_name: str
    reward_rarity: str
    price_paid: int
    created_at: str


# --- admin CRUD ---------------------------------------------------------

class ChestRarityProbabilityIn(BaseModel):
    rarity: str
    probability: float


class ChestCreate(BaseModel):
    slug: str
    name: str
    description: str = ""
    price: int
    image_path: str | None = None
    guaranteed_min_rarity: str | None = None
    is_active: bool = True
    sort_order: int = 0
    rarity_probabilities: list[ChestRarityProbabilityIn]


class ChestUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    price: int | None = None
    image_path: str | None = None
    guaranteed_min_rarity: str | None = None
    is_active: bool | None = None
    sort_order: int | None = None
    rarity_probabilities: list[ChestRarityProbabilityIn] | None = None

File: rpg-frontend/src/types/chest.ts
import type { Rarity } from "./catalog";
import type { ItemAffixOut, ItemStatsOut } from "./item";

export interface ChestRarityProbabilityOut {
  rarity: Rarity;
  probability: number;
}

export interface ChestOut {
  id: number;
  slug: string;
  name: string;
  description: string;
  price: number;
  image_path: string | null;
  guaranteed_min_rarity: Rarity | null;
  is_active: boolean;
  rarity_probabilities: ChestRarityProbabilityOut[];
}

export interface ChestSummaryOut {
  id: number;
  name: string;
}

export interface ChestRewardOut {
  item_id: number;
  item_template_id: number;
  name: string;
  slot: string;
  tier: number;
  rarity: Rarity;
  image_path: string | null;
  stats: ItemStatsOut;
  affixes: ItemAffixOut[];
}

export interface ChestOpenResult {
  opening_id: number;
  chest: ChestSummaryOut;
  reward: ChestRewardOut;
  balance: number;
}

export interface FreeChestStatusOut {
  chest: ChestOut;
  is_available: boolean;
  next_available_at: string | null;
}

export interface ChestOpeningHistoryOut {
  id: number;
  chest: ChestSummaryOut;
  reward_item_id: number;
  reward_item_name: string;
  reward_rarity: Rarity;
  price_paid: number;
  created_at: string;
}

// --- admin CRUD (rpg-backend's admin_chests.py — the only admin-write
// resource that exists today; see FRONTEND_API_MAP.md) ---------------------

export interface ChestRarityProbabilityIn {
  rarity: Rarity;
  probability: number;
}

export interface ChestCreate {
  slug: string;
  name: string;
  description?: string;
  price: number;
  image_path?: string | null;
  guaranteed_min_rarity?: Rarity | null;
  is_active?: boolean;
  sort_order?: number;
  rarity_probabilities: ChestRarityProbabilityIn[];
}

export interface ChestUpdate {
  name?: string;
  description?: string;
  price?: number;
  image_path?: string | null;
  guaranteed_min_rarity?: Rarity | null;
  is_active?: boolean;
  sort_order?: number;
  rarity_probabilities?: ChestRarityProbabilityIn[];
}
